app/models/user.rb

```diff
@@ -0,0 +1,58 @@
+# == Schema Information
+#
+# Table name: users
+#
+#  id              :bigint           not null, primary key
+#  email           :string           not null
+#  name            :string           not null
+#  password_digest :string           not null
+#  session_token   :string           not null
+#  created_at      :datetime         not null
+#  updated_at      :datetime         not null
+#
+# Indexes
+#
+#  index_users_on_session_token  (session_token) UNIQUE
+#
+class User < ApplicationRecord
+    has_secure_password
+
+    validates :name, presence: true, length: { in: 3..30 }
+
+    validates :email, 
+    uniqueness: true, 
+    length: { in: 3..255 }, 
+    format: { with: URI::MailTo::EMAIL_REGEXP }
+
+    validates :session_token, presence: true, uniqueness: true
+    validates :password, length: { in: 6..255 }, allow_nil: true
+  
+    before_validation :ensure_session_token
+  
+    # Ensure session token is set before validation
+
+    def self.find_by_credentials(credential, password)
+        user = User.find_by(email: credential)
```

Comment:
I think here there is no need for `User.find_by` you can do: `user = find_by(email: credentials)`